File: fixtures/bindings-internal/tests/bindings/test_bindings_internal.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import bindings_internal

var slab = UniffiSlab<Int>(0)
let handle1 = try! slab.insert(0)
let handle2 = try! slab.insert(1)
assert(try! slab.get(handle1) == 0)
assert(try! slab.get(handle2) == 1)
let _ = try! slab.remove(handle1)

do {
    let _ = try slab.get(handle1)
    fatalError("get with a removed handle should fail")
} catch UniffiInternalError.slabUseAfterFree {
    // Expected
}

var slab2 = UniffiSlab<Int>(1)
do {
    let _ = try slab2.get(handle2)
    fatalError("get with a handle from a different slab should fail")
} catch UniffiInternalError.slabError {
    // Expected
}

do {
    let _ = try slab.get(handle2 & ~0x0001_0000_0000)
    fatalError("get with a handle from a Rust slab should fail")
} catch UniffiInternalError.slabError {
    // Expected
}

File: fixtures/bindings-internal/tests/bindings/test_bindings_internal.rb
# frozen_string_literal: true

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

require 'test/unit'
require 'bindings_internal'

class TestSlab < Test::Unit::TestCase
  def test_slaB
    slab = BindingsInternal::UniffiSlab.new(0)
    handle1 = slab.insert(0)
    handle2 = slab.insert(1)
    assert_equal(slab.get(handle1), 0)
    assert_equal(slab.get(handle2), 1)
    slab.remove(handle1)
    # Re-using a removed handle should fail
    begin
      slab.get(handle1)
    rescue BindingsInternal::InternalError => err
      # Expected
    else
      raise 'should have thrown'
    end

    # Using a handle with a different slab should fail
    slab2 = BindingsInternal::UniffiSlab.new(1)
    begin
      slab2.get(handle2)
    rescue BindingsInternal::InternalError => err
      # Expected
    else
      raise 'should have thrown'
    end

    # Using a handle from Rust should fail
    begin
      slab.get(handle2 & ~0x0001_0000_0000)
    rescue BindingsInternal::InternalError => err
      # Expected
    else
      raise 'should have thrown'
    end
  end
end
